pkg/filesystem/subvolume.go

```diff
@@ -0,0 +1,173 @@
+				fmt.Println(fs.Name, sv.Name, svg.Name, status)
+			}
+		}
+	}
+}
+
+// gets list of filesystem
+func getFileSystem(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace string) []fsStruct {
+	fsList := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "ls", "--format", "json"}, operatorNamespace, clusterNamespace, true, false)
+	fsstruct := unMarshaljson(fsList)
+	if len(fsstruct) == 0 {
+		logging.Fatal(fmt.Errorf("failed to get filesystem"))
+	}
+	return fsstruct
+}
+
+// checkSnapshot checks if there are any snapshots in the subvolume
+func checkSnapshot(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace, fs, sv, svg string) bool {
+
+	snapList := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "snapshot", "ls", fs, sv, svg}, operatorNamespace, clusterNamespace, true, false)
+	snap := unMarshaljson(snapList)
+	if len(snap) == 0 {
+		return false
+	}
+	return true
+
+}
+
+// gets the list of subvolumegroup for the specified filesystem
+func getSubvolumeGroup(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace, clusterNamespace, fs string) []fsStruct {
+	svgList := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolumegroup", "ls", fs, "--format", "json"}, operatorNamespace, clusterNamespace, true, false)
+	subvolg := unMarshaljson(svgList)
+	if len(subvolg) == 0 {
+		logging.Fatal(fmt.Errorf("failed to get subvolumegroup for filesystem %q", fs))
+	}
+	return subvolg
+}
+
+func unMarshaljson(list string) []fsStruct {
+	var unmarshal []fsStruct
+	errg := json.Unmarshal([]byte(list), &unmarshal)
+	if errg != nil {
+		logging.Fatal(errg)
+	}
+
+	return unmarshal
+}
+
+func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subList, fs, svg string) {
+	subvollist := strings.Split(subList, ",")
+	k8sSubvolume := getK8sRefSubvolume(ctx, clientsets)
+	for _, subvolume := range subvollist {
+		check := checkStaleSubvolume(ctx, clientsets, OperatorNamespace, CephClusterNamespace, fs, subvolume, svg, k8sSubvolume)
+		if check {
+			exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvolume, svg}, OperatorNamespace, CephClusterNamespace, true, false)
+			logging.Info("subvolume %q deleted", subvolume)
+		} else {
+			logging.Info("subvolume %q is not stale", subvolume)
+		}
+	}
+}
+
+// checkStaleSubvolume checks if there are any stale subvolume to be deleted
+func checkStaleSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, fs, subvolume, svg string, k8sSubvolume map[string]subVolumeInfo) bool {
+	_, ok := k8sSubvolume[subvolume]
+	if !ok {
+		snapshot := checkSnapshot(ctx, clientsets, OperatorNamespace, CephClusterNamespace, fs, subvolume, svg)
+		if snapshot {
+			logging.Error(fmt.Errorf("subvolume %s has snapshots", subvolume))
```

Comment:
Is this an error because it caused the deletion to fail? Seems like we should return immediately so we don't also print the other error below
```suggestion
			logging.Error(fmt.Errorf("subvolume %s has snapshots", subvolume))
			return false
```

Reply:
it is error because we don't want to delete the subvolumes which will have snapshots attached to it, to avoid data loss.